log4j-api/src/main/java/org/apache/logging/log4j/ScopedContext.java

```diff
@@ -0,0 +1,225 @@
+     * @param key the key.
+     * @param value the value associated with the key.
+     * @param task the Runnable to call.
+     */
+    public static void runWhere(final String key, final Object value, final Runnable task) {
+        provider.newScopedContext(key, value).run(task);
+    }
+
+    /**
+     * Creates a ScopedContext with a single key/value pair and calls a method on a separate Thread.
+     * @param key the key.
+     * @param value the value associated with the key.
+     * @param executorService the ExecutorService to dispatch the work.
+     * @param task the Runnable to call.
+     */
+    public static Future<Void> runWhere(
+            final String key, final Object value, final ExecutorService executorService, final Runnable task) {
+        return provider.newScopedContext(key, value).run(executorService, task);
+    }
+
+    /**
+     * Creates a ScopedContext with a Map of keys and values and calls a method.
+     * @param map the Map.
+     * @param task the Runnable to call.
+     */
+    public static void runWhere(final Map<String, ?> map, final Runnable task) {
+        provider.newScopedContext(map).run(task);
+    }
+
+    /**
+     * Creates a ScopedContext with a single key/value pair and calls a method.
+     * @param key the key.
+     * @param value the value associated with the key.
+     * @param task the Runnable to call.
+     */
+    public static <R> R callWhere(final String key, final Object value, final Callable<R> task) throws Exception {
+        return provider.newScopedContext(key, value).call(task);
+    }
+
+    /**
+     * Creates a ScopedContext with a single key/value pair and calls a method on a separate Thread.
+     * @param key the key.
+     * @param value the value associated with the key.
+     * @param executorService the ExecutorService to dispatch the work.
+     * @param task the Callable to call.
+     */
+    public static <R> Future<R> callWhere(
+            final String key, final Object value, final ExecutorService executorService, final Callable<R> task) {
+        return provider.newScopedContext(key, value).call(executorService, task);
+    }
+
+    /**
+     * Creates a ScopedContext with a Map of keys and values and calls a method.
+     * @param map the Map.
+     * @param task the Runnable to call.
+     */
+    public static <R> R callWhere(final Map<String, ?> map, final Callable<R> task) throws Exception {
+        return provider.newScopedContext(map).call(task);
+    }
+
+    /**
+     * Return the object with the specified key from the current context.
+     * @param key the key.
+     * @return the value of the key or null.
+     * @param <T> The type of object expected.
+     * @throws ClassCastException if the specified type does not match the object stored.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> T get(String key) {
+        return (T) provider.getValue(key);
+    }
```

Comment:
This method needs to be in the API. Otherwise it will drastically reduce the usage of the class.  As far as OpenTelemetry goes, it is solving a different problem then we are so I am not willing to base decisions we make on what they have decided to do. 

As far as the signature goes, you are correct. I will change it to return Object, which is what the users are storing. That makes it slightly less convenient but they can provide the sugar to make it prettier.

Reply:
> This method needs to be in the API. Otherwise it will drastically reduce the usage of the class. As far as OpenTelemetry goes, it is solving a different problem then we are so I am not willing to base decisions we make on what they have decided to do.

I don't base my decision on OTel, but I happen to agree with them. Can you give a logging-related use case, where a user needs to retrieve a key he previously stored in the context map?

I know our opinions differ, but binary compatibility makes the decision of including `get` or not asymmetrical: if we remove it, we can add it in `2.25.0`; if we add it now, we can not remove it.

Reply:
I have two examples:

1. https://github.com/apache/logging-log4j-audit/blob/master/log4j-audit/log4j-audit-api/src/main/java/org/apache/logging/log4j/audit/rest/RequestContextHeaderInterceptor.java - This supports the ThreadContext. I would need to modify it to support the ScopedContext.
2. We frequently include the tenant id in the ThreadContextMap. We then might have an endpoint that is /dosomething/current.  That means that instead of passing the tenant id in the url the endpoint should retrieve it from the ThreadContextMap. If ScopedContext is used instead then it would need to be able to get the value. Likewise, we may use a datasource or some other resource based on the tenantId.  We do NOT pass the tenant id as a parameter through each and every call in the system so we need to be able to retrieve it. 

Again, everyone who has requested that the MDC support objects wants that so they can retrieve them. There is no other logical reason they would want that since logging can only log them as strings.

Reply:
> We frequently include the tenant id in the ThreadContextMap. We then might have an endpoint that is /dosomething/current. That means that instead of passing the tenant id in the url the endpoint should retrieve it from the ThreadContextMap.

@rgoers, with all due respect, this is a very bad idea. Your application is (ab)using the logging framework for context propagation in between its components. What is your application gonna use `ScopedContext` for next? Passing the total cost while issuing invoices? What is Log4j gonna support next? Introducing `SystemContext` as an in memory key value store?

I understand the use case and I agree that `ScopedContext` of Log4j _can_ help there, but it doesn't mean that it should. MDC has a getter and that was a mistake. If you need a [scoped] context propagation mechanism, I would suggest implementing it as a separate library engineered for that particular use case. I am in favor of only dealing with logging in Log4j and nothing else.

Reply:
@vy with all due respect, get off your high horse. Applications have been using the MDC this way for years and will also use the ScopedContext to do the same, only in a safer manner. They (and I) are not going to stop doing it just because you say so. You are about 20 years too late to argue against the MDC. Arguing against a get method in ScopedContext for this reason is simply silly because all you are accomplishing is saying that users have to continue using something that requires much more care to use safely.

Reply:
Most of the questions about the lack of `Object` support in `ThreadContext` I have seen were logging-related (e.g. [this SO question](https://stackoverflow.com/q/77876650/11748454)).

In those cases users **don't** need to retrieve the object value they put in the thread context, but they want to see an appropriate JSON representation of the value.

Reply:
@ppkarwasz We could go round and round on this for days. The fact of the matter is the get method is required to be able to avoid having to copy the map regardless of whether it is an object or a map. Being able to support objects is just a bonus. However, I would NEVER recommend that the object be converted to JSON on every logging call. The JSON String should be cached - which is one of the reasons I wanted the Renderable interface as toString wouldn't obviously do that. 

As for how MDC data gets used outside of logging - in this link you will find many examples - https://www.tabnine.com/code/java/methods/org.slf4j.MDC/get

Reply:
@rgoers,

Are we talking about `ScopedContext#get` or `ScopedContextProvider#getValue`? The latter is **required** by many logging-related components, the former isn't! The only reason the method is there is to allow users that abused `MDC` as general-purpose key-value store, then did the same thing with `ThreadContext` to continue this dubious practice with `ScopedContext`.

Reply:
I knew you were going to say that. However, I simply don't consider it a dubious practice and, in fact, recommend it. https://logging.apache.org/log4j-audit/latest/requestContext.html. I simply see no point in requiring users who are passing data from client to server to have to store the object in a ThreadContext or ScopedContext and also store it in their own ThreadLocal. It simply makes no sense.